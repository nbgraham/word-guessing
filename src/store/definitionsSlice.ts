import { createAsyncThunk, createSlice } from "@reduxjs/toolkit";
import { Loader } from "../utilities/types";
import { datamuseApi } from "../services/datamuse-api";

export const fetchDefinition = createAsyncThunk(
  "definition/fetchDefinition",
  async (word: string) => {
    const definitions = await datamuseApi.getDefinitions(word);
    return definitions;
  }
);

const definitionsSlice = createSlice({
  name: "definitions",
  initialState: {} as Partial<Record<string, Loader<string[]>>>,
  reducers: {},
  extraReducers: (builder) => {
    builder
      .addCase(fetchDefinition.pending, (state, action) => {
        const word = action.meta.arg;
        state[word] = {
          state: "loading",
        };
      })
      .addCase(fetchDefinition.fulfilled, (state, action) => {
        const word = action.meta.arg;
        const results = action.payload;
        state[word] = {
          state: "done",
          value: results,
        };
      })
      .addCase(fetchDefinition.rejected, (state, action) => {
        const word = action.meta.arg;
        state[word] = {
          state: "error",
        };
      });
  },
});

export default definitionsSlice;
