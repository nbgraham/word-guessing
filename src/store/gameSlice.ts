import { createAsyncThunk, createSlice, PayloadAction } from '@reduxjs/toolkit'
import { unique } from '../utilities/array'
import { evaluateGuess } from '../utilities/guess'
import { WordStatus, AnswerInfo, Loader } from '../utilities/types'
import { WordBank, isAWord, getWordBank } from '../utilities/word-service'

type GameState = {
  answers: Record<
    string,
    {
      guesses: WordStatus[]
      eliminatedLetters: string[]
      foundLetters: string[]
      won?: boolean
    }
  >
  wordBank?: WordBank
  newAnswerInfo: Loader<AnswerInfo>
  answer?: string
}

const initialState: GameState = {
  answers: {},
  newAnswerInfo: {
    state: 'initial',
  },
}

const gameSlice = createSlice({
  name: 'game',
  initialState,
  reducers: {
    submitGuess(
      state,
      action: PayloadAction<{ guess: string; answer: string }>
    ) {
      const { guess, answer } = action.payload
      const answerState = (state.answers[answer] ??= {
        guesses: [],
        eliminatedLetters: [],
        foundLetters: [],
      })
      if (!answerState)
        throw new Error('No entry in state for the requested answer')

      const { eliminatedLetters, foundLetters, status } = evaluateGuess(
        guess,
        answer
      )

      answerState.won = guess.toUpperCase() === answer.toUpperCase()
      answerState.guesses.push(status)
      answerState.eliminatedLetters = unique([
        ...answerState.eliminatedLetters,
        ...eliminatedLetters,
      ])
      answerState.foundLetters = unique([
        ...answerState.foundLetters,
        ...foundLetters,
      ])
    },
    startGame(state, action: PayloadAction<AnswerInfo>) {
      const { wordBank } = state
      const answerInfo = action.payload
      state.answer =
        answerInfo && wordBank && wordBank.version === answerInfo.wordBankId
          ? wordBank.words[answerInfo.answerId].toUpperCase()
          : undefined
      state.newAnswerInfo = { state: 'initial' }
    },
  },
  extraReducers: (builder) => {
    builder
      .addCase(pickNewAnswer.fulfilled, (state, action) => {
        state.newAnswerInfo = {
          state: 'done',
          value: action.payload,
        }
      })
      .addCase(pickNewAnswer.rejected, (state, action) => {
        state.newAnswerInfo = {
          state: 'error',
        }
      })

    builder.addCase(fetchWordBank.fulfilled, (state, action) => {
      state.wordBank = action.payload
    })
  },
})
export default gameSlice

export const pickNewAnswer = createAsyncThunk(
  'game/pickNewAnswer',
  async ({
    mustBeValidWord,
    wordBank,
  }: {
    mustBeValidWord: boolean
    wordBank: WordBank
  }) => {
    let tries = 0
    while (tries < 10) {
      tries++
      const words = wordBank.words
      const index = Math.floor(words.length * Math.random())
      const word = words[index]

      if (!mustBeValidWord || (await isAWord(word))) {
        return {
          answerId: index,
          wordBankId: wordBank?.version,
        } as AnswerInfo
      } else {
        console.warn(`"${word}" is not a word. Looking for another word.`)
      }
    }
    throw new Error('Ran out of tries to pick a new answer')
  }
)

export const fetchWordBank = createAsyncThunk('game/fetchWordBank', () =>
  getWordBank()
)
