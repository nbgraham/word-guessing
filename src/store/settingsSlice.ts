import { createSlice, PayloadAction } from "@reduxjs/toolkit";

const defaultOnlineAnswerServiceVersion = 3;
const defaultOfflineAnswerServiceVersion = 2;

const settingsSlice = createSlice({
  name: "settings",
  initialState: {
    guessesMustBeValidWords: navigator.onLine,
    playOffline: !navigator.onLine,
    answerServiceVersion: navigator.onLine
      ? defaultOnlineAnswerServiceVersion
      : defaultOfflineAnswerServiceVersion,
  },
  reducers: {
    setPlayOffline(state, action: PayloadAction<boolean>) {
      const playOffline = action.payload;
      if (state.playOffline !== playOffline) {
        if (playOffline) {
          state.guessesMustBeValidWords = false;
          state.answerServiceVersion = defaultOfflineAnswerServiceVersion;
        }
      }
      state.playOffline = playOffline;
    },
    setGuessesMustBeValidWords(state, action: PayloadAction<boolean>) {
      state.guessesMustBeValidWords = action.payload;
    },
  },
});
export default settingsSlice;
