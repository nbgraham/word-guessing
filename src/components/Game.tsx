import React, { useCallback } from "react";
import { useAppDispatch, useAppSelector } from "../store";
import Victory from "./Victory";
import gameSlice from "../store/gameSlice";
import { datamuseApi } from "../services/datamuse-api";
import Guess from "./Guess";
import { chooseBestGuess } from "../utilities/guess";
import WordResult from "./WordResult";

const EMPTY: [] = [];

const Game: React.FC<{
  answer: string;
}> = ({ answer }) => {
  const answerState = useAppSelector((state) => state.game.answers[answer]);
  const guesses = answerState?.guesses ?? EMPTY;
  const eliminatedLetters = answerState?.eliminatedLetters ?? EMPTY;
  const foundLetters = answerState?.foundLetters ?? EMPTY;
  const won = answerState?.won ?? false;

  const dispatch = useAppDispatch();
  const handleSubmitGuess = useCallback(
    (guess: string) => {
      dispatch(
        gameSlice.actions.submitGuess({
          guess,
          answer,
        })
      );
    },
    [dispatch, answer]
  );

  const autoGuess = useCallback(() => {
    const spelledLike = new Array(5)
      .fill(null)
      .map((_, i) => {
        const correctLetter = guesses.find(
          (guess) => guess[i].inPosition && guess[i].inWord
        );
        return correctLetter?.[i].character ?? "?";
      })
      .join("");

    return datamuseApi
      .getWordsInfo({
        spelledLike: spelledLike,
        metadata: {
          frequency: true,
        },
      })
      .then((wordsInfo) => {
        const pastGuesses = guesses.map((guess) =>
          guess.map((status) => status.character).join("")
        );

        const bestGuess = chooseBestGuess({
          wordsInfo,
          pastGuesses,
          eliminatedLetters,
          foundLetters,
        });

        if (bestGuess) {
          handleSubmitGuess(bestGuess);
        }
      });
  }, [guesses, eliminatedLetters, foundLetters, handleSubmitGuess]);

  return (
    <div
      style={{
        display: "flex",
        flexDirection: "column",
        alignItems: "center",
        gap: 10,
        paddingTop: 20,
      }}
    >
      {!won && (
        <div>
          <button onClick={() => handleSubmitGuess(answer)}>
            Just tell me the answer
          </button>
        </div>
      )}
      {guesses.map((guess, index) => (
        <WordResult key={index} guessStatus={guess} />
      ))}
      {won ? (
        <Victory guesses={guesses} answer={answer} />
      ) : (
        <>
          <button onClick={autoGuess}>Guess for me</button>
          <Guess
            key={guesses.length}
            onSubmitGuess={handleSubmitGuess}
            eliminatedLetters={eliminatedLetters}
            foundLetters={foundLetters}
          />
        </>
      )}
    </div>
  );
};

export default Game;
