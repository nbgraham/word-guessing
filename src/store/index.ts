import { configureStore } from "@reduxjs/toolkit";
import { useDispatch, useSelector, TypedUseSelectorHook } from "react-redux";
import definitionsSlice from "./definitionsSlice";
import gameSlice from "./gameSlice";
import settingsSlice from "./settingsSlice";

const store = configureStore({
  reducer: {
    game: gameSlice.reducer,
    settings: settingsSlice.reducer,
    definition: definitionsSlice.reducer,
  },
});
export default store;

export type RootState = ReturnType<typeof store.getState>;
export type AppDispatch = typeof store.dispatch;

export const useAppDispatch = () => useDispatch<AppDispatch>();
export const useAppSelector: TypedUseSelectorHook<RootState> = useSelector;
