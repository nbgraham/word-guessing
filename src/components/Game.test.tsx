import { Provider } from "react-redux";
import TestRenderer from "react-test-renderer";
import store from "../store";
import Game from "./Game";

it("renders", () => {
  const tree = TestRenderer.create(
    <Provider store={store}>
      <Game answer={"HEART"} />
    </Provider>
  ).toJSON();
  expect(tree).toMatchSnapshot();
});
