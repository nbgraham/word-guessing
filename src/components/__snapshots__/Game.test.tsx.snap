// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders 1`] = `
<div
  style={
    Object {
      "alignItems": "center",
      "display": "flex",
      "flexDirection": "column",
      "gap": 10,
      "paddingTop": 20,
    }
  }
>
  <div>
    <button
      onClick={[Function]}
    >
      Just tell me the answer
    </button>
  </div>
  <button
    onClick={[Function]}
  >
    Guess for me
  </button>
  <form
    onSubmit={[Function]}
  >
    <input
      disabled={false}
      maxLength={5}
      minLength={5}
      onChange={[Function]}
      pattern="^[a-zA-Z]*$"
      placeholder="Guess"
      required={true}
      title="Only letters are allowed"
      type="text"
      value=""
    />
    <button
      disabled={false}
      type="submit"
    >
      Submit
    </button>
  </form>
  <div
    style={
      Object {
        "display": "flex",
        "flexDirection": "column",
        "gap": 2,
        "maxWidth": 400,
        "paddingTop": 20,
        "width": "100vw",
      }
    }
  >
    <div
      style={
        Object {
          "display": "flex",
          "gap": 4,
          "justifyContent": "center",
        }
      }
    >
      <button
        onClick={[Function]}
        style={
          Object {
            "backgroundColor": "#808080",
            "borderRadius": 4,
            "cursor": "pointer",
            "display": "flex",
            "flex": 1,
            "justifyContent": "center",
            "padding": "15px 0px",
          }
        }
      >
        Q
      </button>
      <button
        onClick={[Function]}
        style={
          Object {
            "backgroundColor": "#808080",
            "borderRadius": 4,
            "cursor": "pointer",
            "display": "flex",
            "flex": 1,
            "justifyContent": "center",
            "padding": "15px 0px",
          }
        }
      >
        W
      </button>
      <button
        onClick={[Function]}
        style={
          Object {
            "backgroundColor": "#808080",
            "borderRadius": 4,
            "cursor": "pointer",
            "display": "flex",
            "flex": 1,
            "justifyContent": "center",
            "padding": "15px 0px",
          }
        }
      >
        E
      </button>
      <button
        onClick={[Function]}
        style={
          Object {
            "backgroundColor": "#808080",
            "borderRadius": 4,
            "cursor": "pointer",
            "display": "flex",
            "flex": 1,
            "justifyContent": "center",
            "padding": "15px 0px",
          }
        }
      >
        R
      </button>
      <button
        onClick={[Function]}
        style={
          Object {
            "backgroundColor": "#808080",
            "borderRadius": 4,
            "cursor": "pointer",
            "display": "flex",
            "flex": 1,
            "justifyContent": "center",
            "padding": "15px 0px",
          }
        }
      >
        T
      </button>
      <button
        onClick={[Function]}
        style={
          Object {
            "backgroundColor": "#808080",
            "borderRadius": 4,
            "cursor": "pointer",
            "display": "flex",
            "flex": 1,
            "justifyContent": "center",
            "padding": "15px 0px",
          }
        }
      >
        Y
      </button>
      <button
        onClick={[Function]}
        style={
          Object {
            "backgroundColor": "#808080",
            "borderRadius": 4,
            "cursor": "pointer",
            "display": "flex",
            "flex": 1,
            "justifyContent": "center",
            "padding": "15px 0px",
          }
        }
      >
        U
      </button>
      <button
        onClick={[Function]}
        style={
          Object {
            "backgroundColor": "#808080",
            "borderRadius": 4,
            "cursor": "pointer",
            "display": "flex",
            "flex": 1,
            "justifyContent": "center",
            "padding": "15px 0px",
          }
        }
      >
        I
      </button>
      <button
        onClick={[Function]}
        style={
          Object {
            "backgroundColor": "#808080",
            "borderRadius": 4,
            "cursor": "pointer",
            "display": "flex",
            "flex": 1,
            "justifyContent": "center",
            "padding": "15px 0px",
          }
        }
      >
        O
      </button>
      <button
        onClick={[Function]}
        style={
          Object {
            "backgroundColor": "#808080",
            "borderRadius": 4,
            "cursor": "pointer",
            "display": "flex",
            "flex": 1,
            "justifyContent": "center",
            "padding": "15px 0px",
          }
        }
      >
        P
      </button>
    </div>
    <div
      style={
        Object {
          "display": "flex",
          "gap": 4,
          "justifyContent": "center",
        }
      }
    >
      <button
        onClick={[Function]}
        style={
          Object {
            "backgroundColor": "#808080",
            "borderRadius": 4,
            "cursor": "pointer",
            "display": "flex",
            "flex": 1,
            "justifyContent": "center",
            "padding": "15px 0px",
          }
        }
      >
        A
      </button>
      <button
        onClick={[Function]}
        style={
          Object {
            "backgroundColor": "#808080",
            "borderRadius": 4,
            "cursor": "pointer",
            "display": "flex",
            "flex": 1,
            "justifyContent": "center",
            "padding": "15px 0px",
          }
        }
      >
        S
      </button>
      <button
        onClick={[Function]}
        style={
          Object {
            "backgroundColor": "#808080",
            "borderRadius": 4,
            "cursor": "pointer",
            "display": "flex",
            "flex": 1,
            "justifyContent": "center",
            "padding": "15px 0px",
          }
        }
      >
        D
      </button>
      <button
        onClick={[Function]}
        style={
          Object {
            "backgroundColor": "#808080",
            "borderRadius": 4,
            "cursor": "pointer",
            "display": "flex",
            "flex": 1,
            "justifyContent": "center",
            "padding": "15px 0px",
          }
        }
      >
        F
      </button>
      <button
        onClick={[Function]}
        style={
          Object {
            "backgroundColor": "#808080",
            "borderRadius": 4,
            "cursor": "pointer",
            "display": "flex",
            "flex": 1,
            "justifyContent": "center",
            "padding": "15px 0px",
          }
        }
      >
        G
      </button>
      <button
        onClick={[Function]}
        style={
          Object {
            "backgroundColor": "#808080",
            "borderRadius": 4,
            "cursor": "pointer",
            "display": "flex",
            "flex": 1,
            "justifyContent": "center",
            "padding": "15px 0px",
          }
        }
      >
        H
      </button>
      <button
        onClick={[Function]}
        style={
          Object {
            "backgroundColor": "#808080",
            "borderRadius": 4,
            "cursor": "pointer",
            "display": "flex",
            "flex": 1,
            "justifyContent": "center",
            "padding": "15px 0px",
          }
        }
      >
        J
      </button>
      <button
        onClick={[Function]}
        style={
          Object {
            "backgroundColor": "#808080",
            "borderRadius": 4,
            "cursor": "pointer",
            "display": "flex",
            "flex": 1,
            "justifyContent": "center",
            "padding": "15px 0px",
          }
        }
      >
        K
      </button>
      <button
        onClick={[Function]}
        style={
          Object {
            "backgroundColor": "#808080",
            "borderRadius": 4,
            "cursor": "pointer",
            "display": "flex",
            "flex": 1,
            "justifyContent": "center",
            "padding": "15px 0px",
          }
        }
      >
        L
      </button>
    </div>
    <div
      style={
        Object {
          "display": "flex",
          "gap": 4,
          "justifyContent": "center",
        }
      }
    >
      <button
        onClick={[Function]}
        style={
          Object {
            "backgroundColor": "#808080",
            "borderRadius": 4,
            "cursor": "pointer",
            "display": "flex",
            "flex": 1,
            "justifyContent": "center",
            "padding": "15px 0px",
          }
        }
      >
        Z
      </button>
      <button
        onClick={[Function]}
        style={
          Object {
            "backgroundColor": "#808080",
            "borderRadius": 4,
            "cursor": "pointer",
            "display": "flex",
            "flex": 1,
            "justifyContent": "center",
            "padding": "15px 0px",
          }
        }
      >
        X
      </button>
      <button
        onClick={[Function]}
        style={
          Object {
            "backgroundColor": "#808080",
            "borderRadius": 4,
            "cursor": "pointer",
            "display": "flex",
            "flex": 1,
            "justifyContent": "center",
            "padding": "15px 0px",
          }
        }
      >
        C
      </button>
      <button
        onClick={[Function]}
        style={
          Object {
            "backgroundColor": "#808080",
            "borderRadius": 4,
            "cursor": "pointer",
            "display": "flex",
            "flex": 1,
            "justifyContent": "center",
            "padding": "15px 0px",
          }
        }
      >
        V
      </button>
      <button
        onClick={[Function]}
        style={
          Object {
            "backgroundColor": "#808080",
            "borderRadius": 4,
            "cursor": "pointer",
            "display": "flex",
            "flex": 1,
            "justifyContent": "center",
            "padding": "15px 0px",
          }
        }
      >
        B
      </button>
      <button
        onClick={[Function]}
        style={
          Object {
            "backgroundColor": "#808080",
            "borderRadius": 4,
            "cursor": "pointer",
            "display": "flex",
            "flex": 1,
            "justifyContent": "center",
            "padding": "15px 0px",
          }
        }
      >
        N
      </button>
      <button
        onClick={[Function]}
        style={
          Object {
            "backgroundColor": "#808080",
            "borderRadius": 4,
            "cursor": "pointer",
            "display": "flex",
            "flex": 1,
            "justifyContent": "center",
            "padding": "15px 0px",
          }
        }
      >
        M
      </button>
      <button
        onClick={[Function]}
        style={
          Object {
            "backgroundColor": "#808080",
            "borderRadius": 4,
            "cursor": "pointer",
            "display": "flex",
            "flex": 1,
            "justifyContent": "center",
            "padding": "15px 0px",
          }
        }
      >
        &lt;
      </button>
      <button
        onClick={[Function]}
        style={
          Object {
            "backgroundColor": "#808080",
            "borderRadius": 4,
            "cursor": "pointer",
            "display": "flex",
            "flex": 1,
            "justifyContent": "center",
            "padding": "15px 0px",
          }
        }
      >
        =
      </button>
    </div>
  </div>
</div>
`;
